trillian/migrillian/election/etcd/election.go

```diff
@@ -0,0 +1,166 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package etcd provides an implementation of master election based on etcd.
+package etcd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/golang/glog"
+	"github.com/google/certificate-transparency-go/trillian/migrillian/election"
+)
+
+var (
+	errMasterAlready     = errors.New("already master")
+	errMasterUnconfirmed = errors.New("mastership unconfirmed")
+	errMasterOvertaken   = errors.New("mastership overtaken")
+)
+
+// Election is an implementation of election.Election based on etcd.
+type Election struct {
+	treeID     int64
+	instanceID string
+	lockFile   string
+
+	client   *clientv3.Client
+	session  *concurrency.Session
+	election *concurrency.Election
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// Await blocks until the instance captures mastership. Returns a "mastership
+// context" which remains active until the instance stops being the master, or
+// the passed in context is canceled. Returns an error if capturing fails, or
+// the passed in context is canceled before mastership is captured.
+func (e *Election) Await(ctx context.Context) (context.Context, error) {
+	// Return an error if the Election already maintains mastership context.
+	if e.ctx != nil && e.ctx.Err() == nil {
+		return nil, errMasterAlready
+	}
+
+	if err := e.election.Campaign(ctx, e.instanceID); err != nil {
+		return nil, err
+	}
+
+	// Get a channel for notifications of election status (using the cancelable
+	// context so that the monitoring goroutine below and the goroutine started
+	// by Observe will reliably terminate).
+	cctx, cancel := context.WithCancel(ctx)
+	ch := e.election.Observe(cctx)
+
+	select {
+	case <-ctx.Done():
+		cancel()
+		return nil, ctx.Err()
+	case rsp, ok := <-ch:
+		if !ok {
+			cancel()
+			return nil, errMasterUnconfirmed
+		}
+		if string(rsp.Kvs[0].Value) != e.instanceID {
+			cancel()
+			return nil, errMasterOvertaken
+		}
+	}
+
+	// At this point we have observed confirmation that we are the master; start
+	// a goroutine to monitor for anyone else overtaking us.
+	go func() {
+		for rsp := range ch {
+			if string(rsp.Kvs[0].Value) != e.instanceID {
+				glog.Warningf("%d: mastership overtaken", e.treeID)
+				break
+			}
+		}
+		glog.Infof("%d: canceling mastership context", e.treeID)
+		cancel()
```

Comment:
self-nit: Better use `defer` in such cases, because what if something in between `panic`s.